service/notification/NotificationReceiver.php

```diff
@@ -0,0 +1,391 @@
+ * #############  #####( ######  #####. ######  #############  #############
+ *        ######  #####( ######  #####. ######  #####  ######  #####  ######
+ * ###### ######  #####( ######  #####. ######  #####  #####   #####  ######
+ * ###### ######  #####( ######  #####. ######  #####          #####  ######
+ * #############  #############  #############  #############  #####  ######
+ *  ############   ############  #############   ############  #####  ######
+ *                                      ######
+ *                               #############
+ *                               ############
+ *
+ * Adyen PrestaShop plugin
+ *
+ * Copyright (c) 2019 Adyen B.V.
+ * This file is open source and available under the MIT license.
+ * See the LICENSE file for more info.
+ */
+
+namespace Adyen\PrestaShop\service\notification;
+
+use Adyen\PrestaShop\helper\Data as AdyenHelper;
+use Adyen\Util\HmacSignature;
+use DateTime;
+use Db;
+
+class NotificationReceiver
+{
+    /**
+     * @var AdyenHelper
+     */
+    private $helperData;
+
+    /**
+     * @var string
+     */
+    private $notificationHMAC;
+
+    /**
+     * @var HmacSignature
+     */
+    private $hmacSignature;
+
+    /**
+     * @var string
+     */
+    private $merchantAccount;
+
+    /**
+     * @var string
+     */
+    private $notificationUsername;
+
+    /**
+     * @var string
+     */
+    private $notificationPassword;
+
+    /**
+     * @var Db
+     */
+    private $dbInstance;
+
+    /**
+     * NotificationReceiver constructor.
+     * @param AdyenHelper $helperData
+     * @param HmacSignature $hmacSignature
+     * @param $notificationHMAC
+     * @param $merchantAccount
+     * @param $notificationUsername
+     * @param $notificationPassword
+     * @param Db $dbInstance
+     */
+    public function __construct(
+        AdyenHelper $helperData,
+        HmacSignature $hmacSignature,
+        $notificationHMAC,
+        $merchantAccount,
+        $notificationUsername,
+        $notificationPassword,
+        Db $dbInstance
+    ) {
+        $this->helperData = $helperData;
+        $this->hmacSignature = $hmacSignature;
+        $this->notificationHMAC = $notificationHMAC;
+        $this->merchantAccount = $merchantAccount;
+        $this->notificationUsername = $notificationUsername;
+        $this->notificationPassword = $notificationPassword;
+        $this->dbInstance = $dbInstance;
+    }
+
+    /**
+     * @param $notificationItems
+     * @return false|string|null
+     * @throws AuthenticationException
+     * @throws HMACKeyValidationException
+     * @throws MerchantAccountCodeException
+     * @throws \Adyen\AdyenException
+     * @throws \PrestaShopDatabaseException
+     * @throws AuthorizationException
+     */
+    public function doPostProcess($notificationItems)
+    {
+        if (!empty($notificationItems['live']) && $this->validateNotificationMode($notificationItems['live'])) {
```

Comment:
notificationItems can be null since json_decode can return null if decoding went wrong
Maybe we could check at first if it is empty and log the error that the notification is not formatted as expected and consuming it failed during the process